#' @title exampleDFGEODownloadR
#' @description An example dataset with data from 3 GSM IDs, used to in examples
#' for GEODownloadR package functions. Note: This dataset contains more than 3
#' rows as it includes information for all SRR files annotated to each GSM ID,
#' and there are often multiple SRR files associated with a given GSM. As an
#' example, a paired end experiment will have at least 2 SRR files associated
#' with it, representing each of the read ends.
#' @format A data frame with 11 rows and 25 variables:
#' \describe{
#'   \item{\code{GSM_IDs}}{(character) The GSM IDs (given as "GSM" followed by
#'      a numeric value - example: GSM2406903).
#'   }
#'   \item{\code{GSM_ID_URLs}}{(character) Links to the pages for the given
#'      GSM ID on the NCBI website.
#'   }
#'   \item{\code{GSE_IDs}}{(character) All GSE IDs associated with each GSM ID
#'      (comma-delimited).
#'   }
#'   \item{\code{GSE_ID_URLs}}{(character) Links to the pages for each of the
#'      GSE IDs on the NCBI website (comma-delimited).
#'   }
#'   \item{\code{SRX_IDs}}{(character) The SRX IDs associated with each GSM ID.}
#'   \item{\code{SRX_ID_URLs}}{(character) Links to the pages for each of the
#'      SRX IDs on the NCBI website.
#'   }
#'   \item{\code{SRR_IDs}}{(character) The SRR IDs associated with each GSM ID.}
#'   \item{\code{SRR_ID_URLs}}{(character) Links to the pages for each of the
#'      SRR IDs on the ENA website.
#'   }
#'   \item{\code{Total_reads}}{(double) The total number of reads in each
#'      SRR .fastq file (as listed from the given file's associated SRX page).
#'   }
#'   \item{\code{Approximate_total_bases}}{(character) The approximate total
#'      bases in each SRR .fastq file, with a letter listing the metric base
#'      unit (as listed from the given file's associated SRX page).
#'   }
#'   \item{\code{Approximate_file_size}}{(character) The approximate total file
#'      size for each SRR .fastq file, including the base unit (as listed from
#'      the given file's associated SRX page).
#'   }
#'   \item{\code{Upload_date}}{(character) The date each file was uploaded
#'      (as listed from the given file's associated SRX page).
#'   }
#'   \item{\code{Species_name}}{(character) Tthe species name for the
#'      organism which the sequencing data was derived from (as listed from the
#'      given file's associated SRR page).
#'   }
#'   \item{\code{Sequencing_instrument}}{(character) The instrument the
#'      sequencing data was generated by (as listed from the given file's
#'      associated SRR page).
#'   }
#'   \item{\code{Experiment_type}}{(character) The type of experiment the
#'      sequencing data came from (as listed from the given file's associated
#'      SRR page).
#'   }
#'   \item{\code{Sequencing_type}}{(character) The type of sequencing
#'      data, either "PAIRED" or "SINGLE" end (as listed from the given file's
#'      associated SRR page).
#'   }
#'   \item{\code{Fastq_URLs}}{(character) Links to each SRR .fastq file on the
#'      ENA FTP server.
#'   }
#'   \item{\code{Fastq_file_names}}{(character) The name of the SRR .fastq
#'      files.
#'   }
#'   \item{\code{Fastq_file_sizes}}{(double) The exact size of the gzipped
#'      SRR .fastq files.
#'   }
#'   \item{\code{Fastq_file_first_upload_date}}{(character) The date each
#'      SRR .fastq file was first uploaded to the ENA server.
#'   }
#'   \item{\code{Passed_GSE_information}}{(character) Values note whether the
#'      GSM ID URL, as well as the SRX and GSE IDs and URLs associated with the
#'      GSM ID were successfully pulled from the NCBI website. Values are
#'      either "PASS" if information was successfully acquired, or "FAILED" if
#'      the information was not.
#'   }
#'   \item{\code{Passed_SRX_information}}{(character) Values note whether the
#'      SRR IDs and URLs, as well as information about the total reads, bases,
#'      file sizes, upload date, species name, sequencing instrument,
#'      experiment type, and sequencing type were successfully pulled from the
#'      NCBI website. Values are either "PASS" if information was successfully
#'      acquired, or "FAILED" if the information was not.
#'   }
#'   \item{\code{Passed_SRR_file_identification}}{(character) Values note
#'      whether the links and names of SRR .fastq files were successfully
#'      pulled from the ENA website. Values are either "PASS" if information
#'      was successfully acquired, or "FAILED" if the information was not.
#'   }
#'   \item{\code{Passed_SRR_file_information}}{(character) Values note whether
#'      the SRR .fastq file sizes and date of upload were successfully pulled
#'      from the ENA website. Values are either "PASS" if information was
#'      successfully acquired, or "FAILED" if the information was not.
#'   }
#'   \item{\code{SRR_file_download_status}}{(character) Values note whether the
#'      SRR .fastq files were successfully downloaded. Values are
#'      "NO_FTP_URL_GIVEN" or "FILE_SIZE_NOT_GIVEN" if prerequisite data was
#'      not available to successfully download, "FAILED_DOWNLOAD" if the file
#'      was not downloaded at all, "FAILED_DOWNLOAD_FILE_TRUNCATED" if a file
#'      was downloaded, but the file size didn't match the expected size for
#'      the file, "GSM_FILE_ALREADY_DOWNLOADED" if a combined GSM file the SRR
#'      file is a component of is already present,
#'      "SRR_FILE_ALREADY_DOWNLOADED" if the SRR .fastq file is already present
#'      and of the correct size, or "PASS" if the file was successfully
#'      downloaded.
#'   }
#'   \item{\code{GSMCombinationStatus}}{(character) Values note whether the
#'      SRR .fastq files were successfully combined into overall .fastq files
#'      annotated to each of the GSM IDs. Values are
#'      "PROPER_SEQUENCING_FORMAT_NOT_RECOGNIZED" if the Sequencing_type value
#'      is NA, or doesn't match for all the SRR entries annotated to a single
#'      unique GSM ID, "ALL_SRR_FILES_NOT_PRESENT" if not all the SRR .fastq
#'      files which need to be combined into a given GSM .fastq file are
#'      present in the specified directory, "IMPROPER_FILE_PAIRING_DETECTED"
#'      where an incorrect number of SRR .fastq files are present for a given
#'      sequencing type (i.e. an odd number of SRR .fastqs are present when the
#'      experiment should be paired), "GSM_FASTQ_NOT_CREATED" if there was some
#'      other error which prevented the GSM .fastq file from being created,
#'      "GSM_FASTQ_FILE_ALREADY_PRESENT" if the combined GSM .fastq file is
#'      already present, "PASS_PAIRED_END_1_CREATED" when the first paired end
#'      file for a GSM is created, but the second is already present, and
#'      conversely "PASS_PAIRED_END_2_CREATED" when the second paired end file
#'      is created but the first is already present, and finally "PASS" when the
#'      GSM .fastq file was successfully combined without other notes.
#'   }
#'}
"exampleDFGEODownloadR"
